src/openforms/tests/e2e/test_input_validation.py

```diff
@@ -670,3 +682,135 @@ def test_regex_pattern(self):
+    ) -> None:
+        form = create_form(component)
+
+        with self.subTest("frontend validation"):
+            self._assertFileFrontendValidation(form, ui_files, expected_ui_error)
+
+        with self.subTest("backend validation"):
+            self._assertBackendValidation(form, component["key"], api_value)
+
+    @async_to_sync
+    async def _assertFileFrontendValidation(
+        self, form: Form, ui_files: list[Path], expected_ui_error: str
+    ) -> None:
+        frontend_path = reverse("forms:form-detail", kwargs={"slug": form.slug})
+        url = str(furl(self.live_server_url) / frontend_path)
+
+        async with browser_page() as page:
+            await page.goto(url)
+            await page.get_by_role("button", name="Formulier starten").click()
+
+            async with page.expect_file_chooser() as fc_info:
+                await page.get_by_text("blader").click()
+
+            file_chooser = await fc_info.value
+            await file_chooser.set_files(ui_files)
+            await page.wait_for_load_state("networkidle")
+
+            # try to submit the step which should be invalid, so we expect this to
+            # render the error message.
+            await page.get_by_role("button", name="Volgende").click()
+            await expect(page.get_by_text(expected_ui_error)).to_be_visible()
+
+    def test_required_field(self):
+        component = {
+            "type": "file",
+            "key": "requiredFile",
+            "label": "Required file",
+            "validate": {"required": True},
+            "storage": "url",
+        }
+
+        self.assertFileValidationIsAligned(
+            component,
+            ui_files=[],
+            expected_ui_error="Het verplichte veld Required file is niet ingevuld.",
+            api_value=[],
+        )
+
+    def test_bad_pdf(self):
+        component = {
+            "type": "file",
+            "key": "badPdf",
+            "label": "Bad PDF",
+            "validate": {"required": True},
+            "storage": "url",
+        }
+
+        # The frontend validation will *not* create a TemporaryFileUpload,
+        # as the endpoint will return a 400 because of the bad content type.
+        # For this reason, we use a random UUID for the `api_value`
+
+        self.assertFileValidationIsAligned(
+            component,
+            ui_files=[TEST_FILES / "image-256x256.pdf"],
+            expected_ui_error="Het bestand is geen .pdf.",
+            api_value=[
+                SubmittedFileFactory.build(
+                    type="application/pdf",
+                    url="http://localhost/api/v2/submissions/files/d4c97feb-68f6-4b85-9b78-7a74ee48b072",
+                    data__url="http://localhost/api/v2/submissions/files/d4c97feb-68f6-4b85-9b78-7a74ee48b072",
+                )
+            ],
+        )
+
+        # Make sure the frontend did not create one:
+        self.assertEqual(TemporaryFileUpload.objects.count(), 0)
```

Comment:
slightly more elegant/semantic:

```suggestion
        # Make sure the frontend did not create one:
        self.assertFalse(TemporaryFileUpload.objects.exists())
```